File: src/parser/lexer.py
from enum import Enum, auto
import re

from ..utils.errors import Error

isEOF = lambda string : string == "\0"
isNum = lambda string : bool(re.compile("^[0-9]+$").search(string))
isValidNum = lambda string : bool(re.compile("^\d+(\.\d)?\d*$").search(string))

class Token:
    def __init__(self, value, type, line, col):
        self.value = value
        self.type = type
        self.location = [ line, col ] 

    def __repr__(self):
        return f"Token({self.value}, {self.type}, {self.location})"

class TokenType(Enum):
        # Single character tokens
        OpenParen      = auto(), # Done
        CloseParen     = auto(), # Done
        Comma          = auto(), # Done
        Dot            = auto(), # Done
        Minus          = auto(), # Done
        Plus           = auto(), # Done
        Slash          = auto(), # Done

        # Literals
        Integer        = auto(), # Done
        Float          = auto(), # Done
        Identifier     = auto(), # Done
        L_String       = auto(), # Done

        # One or two character tokens
        Star           = auto(), # Done
        Mod            = auto(), # Done
        Amp            = auto(), # Done
        DoubleColon    = auto(), # Done
        Arrow          = auto(), # Done
        And            = auto(), # Done
        Or             = auto(), # Done
        Not            = auto(), # Done
        NotEqual       = auto(), # Done
        Equal          = auto(), # Done
        EqualEqual     = auto(), # Done
        Greater        = auto(), # Done
        GreaterEqual   = auto(), # Done
        Less           = auto(), # Done
        LessEqual      = auto(), # Done
        
        # Keywords
        OpenEnd        = auto(), # TODO: how am I supposed to lex this? 
        CloseEnd       = auto(), # Done

        If,            = auto(), # Done
        Else,          = auto(), # Done
        BTrue          = auto(), # Done
        BFalse         = auto(), # Done
        Method         = auto(), # Done
        Extern         = auto(), # Done
        For            = auto(), # Done
        Null           = auto(), # Done
        Print          = auto(), # Done
        Return         = auto(), # Done
        While          = auto(), # Done
        Const          = auto(), # Done
        Let            = auto(), # Done
        
        # Datatypes
        Void           = auto(), # Done

        Int8           = auto(), # Done
        Uint8          = auto(), # Done

        Int16          = auto(), # Done
        Uint16         = auto(), # Done

        Int32          = auto(), # Done
        Uint32         = auto(), # Done

        Int64          = auto(), # Done
        Uint64         = auto(), # Done
    
        Float32        = auto(), # Done
        Float64        = auto(), # Done

        Bool           = auto(), # Done
        String         = auto(), # Done

        # Special
        EOF            = auto()  # Done

        def __repr__(self):
            return f"<{self.name}.{self.value}>"

Keywords = {
    "void": TokenType.Void,

    "i8": TokenType.Int8,
    "u8": TokenType.Uint8,

    "i16": TokenType.Int16,
    "u16": TokenType.Uint16,

    "i32": TokenType.Int32,
    "u32": TokenType.Uint32,

    "i64": TokenType.Int64,
    "u64": TokenType.Uint64,
        
    "f32": TokenType.Float32,
    "f64": TokenType.Float64,

    "bool": TokenType.Bool,
    "string": TokenType.String,
    
    "end": TokenType.CloseEnd,
    "if": TokenType.If,
    "else": TokenType.Else,
    "true": TokenType.BTrue,
    "false": TokenType.BFalse,
    "extern": TokenType.Extern,
    "method": TokenType.Method,
    "for": TokenType.For,
    "null": TokenType.Null,
    "print": TokenType.Print,
    "return": TokenType.Return,
    "while": TokenType.While,
    "const": TokenType.Const,
    "let": TokenType.Let
}

class Lexer:
    def __init__(self, path, source):
        self.path = path
        self.source = str(source)
        self.buffer = list(self.source)
        self.buffer.append(" ")
        self.buffer.append("\0") # EOF
        self.tokens = []

        # location
        self.line = 1
        self.col = 1

        self.spaces = ["\r", " ", "\t"]
        self.strings = ["'", '"']
    
    def loc_cpy(self):
        return { "line": self.line, "col": self.col }

    def advance(self):
        if self.buffer[0] == "\n":
            self.line += 1
            self.col = 1
            return self.buffer.pop(0)
        elif isEOF(self.buffer[0]):
            self.buffer.pop(0)
            return self.tokens.append(Token("EOF", TokenType["EOF"], self.line, self.col)) 
        else:
            self.col += 1
            return self.buffer.pop(0)
        
    def eat_whitespaces(self):
        while self.buffer[0] in self.spaces and not isEOF(self.buffer[0]):
            self.advance()

    def make_string(self):
        loc = self.loc_cpy()
        delim = self.advance()
        string_content = "" 
        while not isEOF(self.buffer[0]) and self.buffer[0] != delim:
            string_content += self.advance()

        if self.buffer[0] == delim: self.advance()
        
        if isEOF(self.buffer[0]):
            Error(self.path, loc["line"], loc["col"], "Unexpected EOF", "Expected end of string literal but found end of file.")

        self.tokens.append(Token(f"{delim}{string_content}{delim}", TokenType["L_String"], loc["line"], loc["col"]))

    def make_number(self):
        loc = self.loc_cpy()
        number = ""
        
        while isNum(self.buffer[0]) or self.buffer[0] == ".":
            number += self.advance()
        
        if not isValidNum(number): Error(self.path, loc["line"], loc["col"], "Invalid number", "Invalid number found:", number)

        if "." in number:
            self.tokens.append(Token(number, TokenType["Float"], loc["line"], loc["col"]))
        else:
            self.tokens.append(Token(number, TokenType["Integer"], loc["line"], loc["col"]))

    def make_identifier(self):
        loc = self.loc_cpy()
        ident = ""

        while not isEOF(self.buffer[0]) and self.buffer[0].isalnum():
            ident += self.advance()
        
        reserved = Keywords.get(ident)
        if reserved is None:
            self.tokens.append(Token(ident, TokenType["Identifier"], loc["line"], loc["col"]))
        else:
            self.tokens.append(Token(ident, reserved, loc["line"], loc["col"]))

    def lex(self):
        while len(self.buffer) > 0:
            # Newline
            if self.buffer[0] == "\n":
                self.advance()
            # Spaces    
            elif self.buffer[0] in self.spaces:
                self.eat_whitespaces()
            # OpenParen
            elif self.buffer[0] == "(":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["OpenParen"], loc["line"], loc["col"]))
            # CloseParen
            elif self.buffer[0] == ")":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["CloseParen"], loc["line"], loc["col"]))
            # Dot
            elif self.buffer[0] == ".":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["Dot"], loc["line"], loc["col"]))
            # Comma
            elif self.buffer[0] == ",":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["Comma"], loc["line"], loc["col"]))
            # Minus
            elif self.buffer[0] == "-":
                loc = self.loc_cpy()
                self.advance()
                # Arrow
                if self.buffer[0] == ">":
                    self.advance()
                    self.tokens.append(Token("->", TokenType["Arrow"], loc["line"], loc["col"]))
                else:
                    self.tokens.append(Token("-", TokenType["Minus"], loc["line"], loc["col"]))
            # Plus
            elif self.buffer[0] == "+":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["Plus"], loc["line"], loc["col"]))
            # Equal
            elif self.buffer[0] == "=":
                loc = self.loc_cpy()
                self.advance()
                # Equal Equal
                if self.buffer[0] == "=":
                    self.advance()
                    self.tokens.append(Token("==", TokenType["EqualEqual"], loc["line"], loc["col"]))
                else:  
                    self.tokens.append(Token("=", TokenType["Equal"], loc["line"], loc["col"]))
            elif self.buffer[0] == "!":
                loc = self.loc_cpy()
                self.advance()
                # Not Equal
                if self.buffer[0] == "=":
                    self.advance()
                    self.tokens.append(Token("!=", TokenType["NotEqual"], loc["line"], loc["col"]))
                else:
                    self.tokens.append(Token("!", TokenType["Not"], loc["line"], loc["col"]))
            elif self.buffer[0] == "<":
                loc = self.loc_cpy()
                self.advance()
                # Less Equal
                if self.buffer[0] == "=":
                    self.advance()
                    self.tokens.append(Token("<=", TokenType["LessEqual"], loc["line"], loc["col"]))
                # Less
                else:
                    self.tokens.append(Token("<", TokenType["Less"], loc["line"], loc["col"]))
            elif self.buffer[0] == ">":
                loc = self.loc_cpy()
                self.advance()
                # Greater Equal
                if self.buffer[0] == "=":
                    self.advance()
                    self.tokens.append(Token(">=", TokenType["GreaterEqual"], loc["line"], loc["col"]))
                else:
                    self.tokens.append(Token(">", TokenType["Greater"], loc["line"], loc["col"]))
            elif self.buffer[0] == "/":
                loc = self.loc_cpy()
                self.advance()
                # Single line comment
                if self.buffer[0] == "/":
                    self.advance()
                    while self.buffer[0] != "\n" and not isEOF(self.buffer[0]):
                        self.advance()
                # Multi line comment
                elif self.buffer[0] == "*":
                    self.advance()
                    while ((not isEOF(self.buffer[0])) and (self.buffer[0] != "*" and self.buffer[1] != "/")):
                        self.advance()

                    if isEOF(self.buffer[0]):
                        Error(self.path, loc["line"], loc["col"], "Unexpected EOF", "Expected end of comment but found end of file.")

                    self.advance()
                    self.advance()
                # Slash
                else:
                    self.tokens.append(Token("/", TokenType["Slash"], loc["line"], loc["col"]))
            elif self.buffer[0] == "&":
                loc = self.loc_cpy()
                self.advance()
                # And
                if self.buffer[0] == "&":
                    self.advance()
                    self.tokens.append(Token("&&", TokenType["And"], loc["line"], loc["col"]))
                # Amp
                else:
                    self.tokens.append(Token("&", TokenType["Amp"], loc["line"], loc["col"]))
            # Or
            elif self.buffer[0] == "|" and self.buffer[1] == "|":
                loc = self.loc_cpy()
                self.advance()
                self.advance()
                self.tokens.append(Token("||", TokenType["Or"], loc["line"], loc["col"]))
            # Double colon
            elif self.buffer[0] == ":" and self.buffer[1] == ":":
                loc = self.loc_cpy()
                self.advance()
                self.advance()
                self.tokens.append(Token("::", TokenType["DoubleColon"], loc["line"], loc["col"]))
            # Star
            elif self.buffer[0] == "*":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["Star"], loc["line"], loc["col"]))
            # Modulus
            elif self.buffer[0] == "%":
                loc = self.loc_cpy()
                self.tokens.append(Token(self.advance(), TokenType["Mod"], loc["line"], loc["col"]))
            # L_String
            elif self.buffer[0] in self.strings:
                self.make_string()
            else:
                # Numbers
                if isNum(self.buffer[0]):
                    self.make_number()
                # Identifiers/Keywords
                elif self.buffer[0].isalpha():
                    self.make_identifier()
                elif self.buffer[0] == "\0":
                    self.advance()
                else:
                    loc = self.loc_cpy()
                    Error(self.path, loc["line"], loc["col"], "Illegal character", "Invalid character found:", self.advance())

        return self.tokens

__all__ = [
    "Token",
    "TokenType",
    "Keywords",
    "Lexer"
]

if __name__ == "__main__":
    import os
    raise Exception(f"`{os.path.basename(__file__)} can\'t be ran as `{__name__}`")

File: src/parser/lexer/lexer.js
const { Error } = require("../../utils/errors")
const { Token, TokenType, Keywords } = require("./tokens")

// Checks if string is EOF
const isEOF      = (string) => string == "\0"

// Check if string is a number
const isNum      = (string) => /^[0-9]+$/.test(string)

// Check if strings is a valid number
const isValidNum = (string) => /^\d+(\.\d)?\d*$/.test(string)

// Check if string is alphabetic
const isAlpha    = (string) => /[a-zA-Z]/.test(string)

// Checks if a string is alphanumeric
const isAlnum    = (string) => /^[a-zA-Z0-9]+$/.test(string)

class Lexer {
    constructor(path, source) {
        
        // Path
        this.path = path

        // File-related stuff
        this.source = source
        this.buffer = this.source.split("")
        
        // Buffer initialization
        this.buffer.push(" ")  // Whitespace
        this.buffer.push("\0") // EOF

        this.tokens = []

        // Location
        this.line = 1
        this.col = 1

        // Miscellaneous
        this.spaces = ["\r", " ", "\t"]
        this.strings = ["'", '"']
    }

    // Copy the current location
    loc_cpy() {
        return {
            "line": this.line,
            "col":  this.col
        }
    }
    
    advance() {
        
        // Newline
        if (this.buffer[0] == "\n") {
            this.line += 1
            this.col = 1

            return this.buffer.shift()

        // EOF
        } else if (isEOF(this.buffer[0])) {
            const loc = this.loc_cpy()

            this.buffer.shift()

            return this.tokens.push(new Token("EOF", TokenType.EOF, loc.line, loc.col)) 
        // ...
        } else {
            this.col += 1

            return this.buffer.shift()
        }
    }

    eat_whitespaces() {
        while (this.spaces.some((k) => k == this.buffer[0]) && !isEOF(this.buffer[0])) {
            this.advance()
        }
    }

    make_string() {
        const loc = this.loc_cpy()

        const delim = this.advance()
        let stringContent = "" 

        while (!isEOF(this.buffer[0]) && this.buffer[0] != delim) {
            stringContent += this.advance()
        }

        if (this.buffer[0] == delim) {
            this.advance()
        }
        
        if (isEOF(this.buffer[0])) {
            new Error(this.path, loc.line, loc.col, "Unexpected EOF", "Expected end of string literal but found end of file.")
        }

        this.tokens.push(new Token(`${delim}${stringContent}${delim}`, TokenType.L_String, loc.line, loc.col))
    }

    make_number() {
        const loc = this.loc_cpy()

        let number = ""
        
        while (isNum(this.buffer[0]) || this.buffer[0] == ".") {
            number += this.advance()
        }
        
        if (!isValidNum(number)) {
            new Error(this.path, loc.line, loc.col, "Invalid number", `Invalid number found: \`${number}\``)
        }

        if (number.includes(".")) {
            this.tokens.push(new Token(number, TokenType.Float, loc.line, loc.col))
        } else {
            this.tokens.push(new Token(number, TokenType.Integer, loc.line, loc.col))
        }
    }

    make_identifier() {
        const loc = this.loc_cpy()

        let ident = ""

        while (!isEOF(this.buffer[0]) && isAlnum(this.buffer[0])) {
            ident += this.advance()
        }
        
        const reserved = Keywords[ident] 

        // Identifier
        if (reserved == undefined) {
            this.tokens.push(new Token(ident, TokenType.Identifier, loc.line, loc.col))
        // Keyword
        } else {
            this.tokens.push(new Token(ident, reserved, loc["line"], loc["col"]))
        }
    }

    lex() {

        while (this.buffer.length > 0) {

            // Newline
            if (this.buffer[0] == "\n") {
                this.advance()
            // Spaces 
            } else if (this.spaces.some((k) => this.buffer[0] == k)) {
                this.eat_whitespaces()
            // OpenParen
            } else if (this.buffer[0] == "(") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.OpenParen, loc.line, loc.col))
            // CloseParen
            } else if (this.buffer[0] == ")") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.CloseParen, loc.line, loc.col))
            // Dot
            } else if (this.buffer[0] == ".") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.Dot, loc.line, loc.col))
            // Comma
            } else if (this.buffer[0] == ",") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.Comma, loc.line, loc.col))
            // Minus
            } else if (this.buffer[0] == "-") {
                const loc = this.loc_cpy()

                this.advance()

                // Arrow
                if (this.buffer[0] == ">") {
                    this.advance()

                    this.tokens.push(new Token("->", TokenType.Arrow, loc.line, loc.col))
                } else {
                    this.tokens.push(new Token("-", TokenType.Minus, loc.line, loc.col))
                }
            // Plus
            } else if (this.buffer[0] == "+") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType["Plus"], loc["line"], loc["col"]))
            // Equal
            } else if (this.buffer[0] == "=") {
                const loc = this.loc_cpy()

                this.advance()

                // Equal Equal
                if (this.buffer[0] == "=") {
                    this.advance()
                    this.tokens.push(new Token("==", TokenType.EqualEqual, loc.line, loc.col))
                } else { 
                    this.tokens.push(new Token("=", TokenType.Equal, loc.line, loc.col))
                }
            // Not
            } else if (this.buffer[0] == "!") {
                const loc = this.loc_cpy()

                this.advance()

                // Not Equal
                if (this.buffer[0] == "=") {
                    this.advance()
                    this.tokens.push(new Token("!=", TokenType.NotEqual, loc.line, loc.col))
                } else {
                    this.tokens.push(new Token("!", TokenType.Not, loc.line, loc.col))
                }
            // Less
            } else if (this.buffer[0] == "<") {
                const loc = this.loc_cpy()

                this.advance()

                // Less Equal
                if (this.buffer[0] == "=") {
                    this.advance()

                    this.tokens.push(new Token("<=", TokenType.LessEqual, loc.line, loc.col))
                } else {
                    this.tokens.push(new Token("<", TokenType.Less, loc.line, loc.col))
                }
            // Greater
            } else if (this.buffer[0] == ">") {
                const loc = this.loc_cpy()

                this.advance()

                // Greater Equal
                if (this.buffer[0] == "=") {
                    this.advance()
                    this.tokens.push(new Token(">=", TokenType.GreaterEqual, loc.line, loc.col))
                } else {
                    this.tokens.push(new Token(">", TokenType.Greater, loc.line, loc.col))
                }
            // Slash
            } else if (this.buffer[0] == "/") {
                const loc = this.loc_cpy()

                this.advance()

                // Single line comment
                if (this.buffer[0] == "/") {
                    this.advance()

                    while (this.buffer[0] != "\n" && !isEOF(this.buffer[0])) {
                        this.advance()
                    }
                // Multi line comment
                } else if (this.buffer[0] == "*") {
                    this.advance()

                    while ((!isEOF(this.buffer[0])) && (this.buffer[0] != "*" && this.buffer[1] != "/")) {
                        this.advance()
                    }

                    if (isEOF(this.buffer[0])) {
                        new Error(this.path, loc.line, loc.col, "Unexpected EOF", "Expected end of comment but found end of file.")
                    }

                    this.advance() // Eat `*`
                    this.advance() // Eat `/`
                } else {
                    this.tokens.push(new Token("/", TokenType.Slash, loc.line, loc.col))
                }
            // Amp
            } else if (this.buffer[0] == "&") {
                const loc = this.loc_cpy()

                this.advance()

                // And
                if (this.buffer[0] == "&") {
                    this.advance()

                    this.tokens.push(new Token("&&", TokenType.And, loc.line, loccol))
                } else {
                    this.tokens.push(new Token("&", TokenType.Amp, loc.line, loc.col))
                }
            // Or
            } else if (this.buffer[0] == "|" && this.buffer[1] == "|") {
                const loc = this.loc_cpy()

                this.advance() // Eat `|`
                this.advance() // Eat `|`

                this.tokens.push(new Token("||", TokenType.Or, loc.line, loc.col))
            // Double colon
            } else if (this.buffer[0] == ":" && this.buffer[1] == ":") {
                const loc = this.loc_cpy()

                this.advance() // Eat `:`
                this.advance() // Eat `:`

                this.tokens.push(new Token("::", TokenType.DoubleColon, loc.line, loc.col))
            // Star
            } else if (this.buffer[0] == "*") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.Star, loc.line, loc.col))
            // Modulus
            } else if (this.buffer[0] == "%") {
                const loc = this.loc_cpy()

                this.tokens.push(new Token(this.advance(), TokenType.Mod, loc.line, loc.col))
            // L_String
            } else if (this.strings.some((k) => k == this.buffer[0])) {
                this.make_string()
            } else {

                // Numbers
                if (isNum(this.buffer[0])) {
                    this.make_number()
                // Identifiers / Keywords
                } else if (isAlpha(this.buffer[0])) {
                    this.make_identifier()
                // EOF
                } else if (this.buffer[0] == "\0") {
                    this.advance()
                // Unknown character
                } else {
                    const loc = this.loc_cpy()

                    new Error(this.path, loc.line, loc.col, "Illegal character", `Invalid character found: ${this.advance()}`)
                }
            }
        }

        return this.tokens
    }

}

module.exports = {
    Lexer
}

if (require.main === module) {
    const process = require("process")
    const path = require("path")

    console.error(`ERROR: \`${path.parse(__filename).name}\` file can\'t be ran as main.`)
    process.exit(1)
}